File: ios/Classes/InMobiBannerPlatformView.swift
import Flutter
import InMobiSDK
import UIKit

/// Builds the platform views that back `InMobiBannerAd`.
final class InMobiBannerViewFactory: NSObject, FlutterPlatformViewFactory {

    private let messenger: FlutterBinaryMessenger
    private let sendEvent: (Int, String, [String: Any?]) -> Void

    init(
        messenger: FlutterBinaryMessenger,
        sendEvent: @escaping (Int, String, [String: Any?]) -> Void
    ) {
        self.messenger = messenger
        self.sendEvent = sendEvent
        super.init()
    }

    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }

    func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
        return InMobiBannerPlatformView(
            frame: frame,
            params: args as? [String: Any] ?? [:],
            sendEvent: sendEvent
        )
    }
}

/// One `IMBanner`, embedded in the Flutter view hierarchy.
///
/// iOS takes the banner size from the frame in points, which is the same unit
/// Flutter's logical pixels are — so unlike Android there is no density
/// conversion here. A frame that does not match the creative size the placement
/// serves renders blank rather than scaling.
final class InMobiBannerPlatformView: NSObject, FlutterPlatformView, IMBannerDelegate {

    private let adId: Int
    private let sendEvent: (Int, String, [String: Any?]) -> Void
    private let container: UIView
    private var banner: IMBanner?

    init(
        frame: CGRect,
        params: [String: Any],
        sendEvent: @escaping (Int, String, [String: Any?]) -> Void
    ) {
        self.adId = params["adId"] as? Int ?? -1
        self.sendEvent = sendEvent

        let width = (params["width"] as? NSNumber)?.doubleValue ?? 320
        let height = (params["height"] as? NSNumber)?.doubleValue ?? 50
        let bannerFrame = CGRect(x: 0, y: 0, width: width, height: height)

        self.container = UIView(frame: bannerFrame)
        super.init()

        guard let placementId = (params["placementId"] as? NSNumber)?.int64Value else {
            sendEvent(
                adId,
                "loadFailed",
                ["code": "INTERNAL_ERROR", "message": "placementId is required"]
            )
            return
        }

        let banner = IMBanner(frame: bannerFrame, placementId: placementId)
        banner.delegate = self

        // InMobi clamps anything under 20 seconds up to its floor rather than
        // honouring it. Zero is this package's own signal for "off".
        switch (params["refreshIntervalSeconds"] as? NSNumber)?.intValue {
        case .none:
            break
        case .some(0):
            banner.shouldAutoRefresh(false)
        case .some(let seconds):
            banner.refreshInterval = seconds
        }

        // Banners live inside a Flutter layout that animates on its own terms;
        // InMobi's default transition on refresh fights it.
        banner.transitionAnimation = .none

        self.banner = banner
        container.addSubview(banner)
        banner.load()
    }

    func view() -> UIView {
        return container
    }

    // MARK: - IMBannerDelegate

    func bannerDidFinishLoading(_ banner: IMBanner) {
        sendEvent(adId, "loaded", [:])
    }

    func banner(_ banner: IMBanner, didFailToLoadWithError error: IMRequestStatus) {
        sendEvent(adId, "loadFailed", InMobiStatus.eventMap(error))
    }

    func bannerAdImpressed(_ banner: IMBanner) {
        sendEvent(adId, "impression", [:])
    }

    func banner(_ banner: IMBanner, didInteractWithParams params: [String: Any]?) {
        sendEvent(adId, "clicked", [:])
    }
}

File: android/src/main/kotlin/com/illuminationdevelopment/inmobi_ads/InMobiBannerPlatformView.kt
package com.illuminationdevelopment.inmobi_ads

import android.content.Context
import android.view.View
import android.view.ViewGroup
import com.inmobi.ads.AdMetaInfo
import com.inmobi.ads.InMobiAdRequestStatus
import com.inmobi.ads.InMobiBanner
import com.inmobi.ads.listeners.BannerAdEventListener
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

/** Builds the [InMobiBannerPlatformView]s that back `InMobiBannerAd`. */
internal class InMobiBannerViewFactory(
    private val sendEvent: EventSink,
) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        val params = args as? Map<*, *> ?: emptyMap<String, Any?>()
        return InMobiBannerPlatformView(context, params, sendEvent)
    }
}

/**
 * One `InMobiBanner`, embedded in the Flutter view hierarchy.
 *
 * ### Sizing
 *
 * InMobi's Android banner takes its size from its layout params **in device
 * pixels**, not from a size constant the way AdMob does. Dart sends logical
 * pixels and the conversion happens here, against the density of the context
 * the view is actually attached to — reusing a density Flutter reported earlier
 * puts the banner at the wrong size on a second display or after a fold.
 *
 * A banner whose layout params do not match the creative size the placement
 * serves renders blank rather than scaling, which is the usual cause of a
 * banner that reports `loaded` and shows nothing.
 */
internal class InMobiBannerPlatformView(
    context: Context,
    params: Map<*, *>,
    private val sendEvent: EventSink,
) : PlatformView {

    private val adId = (params["adId"] as? Number)?.toInt() ?: -1
    private val banner: InMobiBanner?

    init {
        val placementId = (params["placementId"] as? Number)?.toLong()
        val density = context.resources.displayMetrics.density
        val widthPx = ((params["width"] as? Number)?.toFloat() ?: 320f) * density
        val heightPx = ((params["height"] as? Number)?.toFloat() ?: 50f) * density

        banner = if (placementId == null) {
            sendEvent(
                adId,
                "loadFailed",
                mapOf("code" to "INTERNAL_ERROR", "message" to "placementId is required"),
            )
            null
        } else {
            InMobiBanner(context, placementId).apply {
                setListener(BannerListener())
                setEnableAutoRefresh(true)

                // InMobi clamps anything under 20 seconds up to its floor rather
                // than honouring it, so a caller asking for 5 gets 20 and no
                // warning. Zero is this package's own signal for "off".
                when (val seconds = (params["refreshIntervalSeconds"] as? Number)?.toInt()) {
                    null -> Unit
                    0 -> setEnableAutoRefresh(false)
                    else -> setRefreshInterval(seconds)
                }

                // Banners live inside a Flutter layout that animates on its own
                // terms; InMobi's default axis rotation on refresh fights it.
                setAnimationType(InMobiBanner.AnimationType.ANIMATION_OFF)

                layoutParams = ViewGroup.LayoutParams(widthPx.toInt(), heightPx.toInt())
                load()
            }
        }
    }

    override fun getView(): View? = banner

    override fun dispose() {
        banner?.destroy()
    }

    private inner class BannerListener : BannerAdEventListener() {
        override fun onAdLoadSucceeded(ad: InMobiBanner, info: AdMetaInfo) =
            sendEvent(adId, "loaded", emptyMap())

        override fun onAdLoadFailed(ad: InMobiBanner, status: InMobiAdRequestStatus) =
            sendEvent(adId, "loadFailed", status.toEventMap())

        override fun onAdImpression(ad: InMobiBanner) =
            sendEvent(adId, "impression", emptyMap())

        override fun onAdClicked(ad: InMobiBanner, params: MutableMap<Any, Any>?) =
            sendEvent(adId, "clicked", emptyMap())
    }
}
